File: src/agentswarm/adapters/runner.py
"""The thin subprocess boundary: run one external CLI, capture everything.

This is the only place in the package that actually spawns an external process.
Everything above it (the bridge, the primitives, the runtime) is expressed in
terms of :class:`CliResult`, which makes the higher layers testable without
real agent accounts — a test either points an adapter at a deterministic mock
command, or injects a fake ``run`` callable with the same signature.
"""

from __future__ import annotations

import subprocess
import time
from dataclasses import dataclass
from typing import Protocol


@dataclass(frozen=True)
class CliResult:
    """The outcome of a single CLI invocation."""

    returncode: int
    stdout: str
    stderr: str
    timed_out: bool
    duration: float

    @property
    def ok(self) -> bool:
        """True when the process exited cleanly and did not time out."""
        return self.returncode == 0 and not self.timed_out


class CommandRunner(Protocol):
    """The injectable contract for executing a command. ``run_command`` fulfils it."""

    def __call__(
        self,
        command: list[str],
        *,
        stdin: str | None = None,
        cwd: str | None = None,
        env: dict[str, str] | None = None,
        timeout: float | None = None,
    ) -> CliResult: ...


def run_command(
    command: list[str],
    *,
    stdin: str | None = None,
    cwd: str | None = None,
    env: dict[str, str] | None = None,
    timeout: float | None = None,
) -> CliResult:
    """Run *command*, returning a :class:`CliResult` instead of raising.

    A timeout or a missing executable is reported through the result
    (``timed_out`` / a non-zero ``returncode`` with the reason on stderr)
    rather than as an exception, so the caller has one uniform thing to inspect.
    """
    started = time.monotonic()
    try:
        completed = subprocess.run(
            command,
            input=stdin,
            cwd=cwd,
            env=env,
            capture_output=True,
            text=True,
            timeout=timeout,
        )
    except subprocess.TimeoutExpired as exc:
        return CliResult(
            returncode=-1,
            stdout=_as_text(exc.stdout),
            stderr=_as_text(exc.stderr),
            timed_out=True,
            duration=time.monotonic() - started,
        )
    except (FileNotFoundError, PermissionError, OSError) as exc:
        return CliResult(
            returncode=127,
            stdout="",
            stderr=f"failed to launch {command[0]!r}: {exc}",
            timed_out=False,
            duration=time.monotonic() - started,
        )
    return CliResult(
        returncode=completed.returncode,
        stdout=completed.stdout or "",
        stderr=completed.stderr or "",
        timed_out=False,
        duration=time.monotonic() - started,
    )


def _as_text(value: str | bytes | None) -> str:
    if value is None:
        return ""
    if isinstance(value, bytes):
        return value.decode("utf-8", errors="replace")
    return value

File: src/adapters/runner.ts
/**
 * The thin subprocess boundary (L1).
 *
 * The only place that actually spawns an external process. Everything above it
 * is expressed in terms of {@link CliResult}, which keeps the higher layers
 * testable without real agent accounts — a test injects a fake runner with the
 * same signature.
 *
 * A timeout or a missing executable is reported *through the result*
 * (`timedOut` / a non-zero `returncode` with the reason on stderr) rather than
 * as a thrown error, so the caller has one uniform thing to inspect.
 */

import { spawn } from "node:child_process";

import type { CliResult } from "./types.js";

export interface RunCommandOptions {
  stdin?: string;
  cwd?: string;
  env?: Record<string, string>;
  /** Seconds before the process is killed; omit for no timeout. */
  timeout?: number;
}

/** The injectable contract for executing a command. */
export type CommandRunner = (command: string[], options?: RunCommandOptions) => Promise<CliResult>;

export const runCommand: CommandRunner = (command, options = {}) => {
  const started = Date.now();
  const elapsed = (): number => (Date.now() - started) / 1000;
  const [cmd, ...args] = command;

  return new Promise<CliResult>((resolve) => {
    if (!cmd) {
      resolve({ returncode: 127, stdout: "", stderr: "empty command", timedOut: false, duration: 0 });
      return;
    }

    const child = spawn(cmd, args, {
      cwd: options.cwd,
      env: options.env ?? (process.env as Record<string, string>),
    });

    let stdout = "";
    let stderr = "";
    let timedOut = false;
    let settled = false;
    let timer: NodeJS.Timeout | undefined;

    const finish = (result: CliResult): void => {
      if (settled) return;
      settled = true;
      if (timer) clearTimeout(timer);
      resolve(result);
    };

    if (options.timeout != null) {
      timer = setTimeout(() => {
        timedOut = true;
        child.kill("SIGKILL");
      }, options.timeout * 1000);
    }

    child.stdout.setEncoding("utf8");
    child.stderr.setEncoding("utf8");
    child.stdout.on("data", (d: string) => {
      stdout += d;
    });
    child.stderr.on("data", (d: string) => {
      stderr += d;
    });

    child.on("error", (err) => {
      finish({
        returncode: 127,
        stdout: "",
        stderr: `failed to launch '${cmd}': ${err.message}`,
        timedOut: false,
        duration: elapsed(),
      });
    });

    child.on("close", (code) => {
      finish({
        returncode: timedOut ? -1 : code ?? 0,
        stdout,
        stderr,
        timedOut,
        duration: elapsed(),
      });
    });

    if (options.stdin != null) child.stdin.write(options.stdin);
    child.stdin.end();
  });
};
